### PR TITLE
[WIP] Tool linting fixes and proper exit_code detection

### PR DESCRIPTION
**Pull request by hexylena**
I'm linting all of the tools in the tool directory with planemo and identifying and fixing issues. I'm opening this PR very early into the process so I can track some discussion questions.

The primary purpose of the PR is to use https://github.com/galaxyproject/galaxy/pull/117 throughout so tools don't die on any stderr, as that's really poor behaviour. This is specifically in response to a `UserWarning` showing up during upload, a warning which causes the job to unnecessarily error.
- [ ] can we nix `import.py` or replace it with a `cp` command?
- [x] removed echo.py
- [ ] there are at least a couple of tools with the same ID 
  - ucsc_proxy for `ucsc_archaea.xml` and `ucsc_proxy.xml`
  - Grep1 for `grep.xml` and `fileGrep.xml`
- [ ] `access_libraries.xml`? Is this still a thing? It seems like an API call, but no idea if it's even functional.
- [x] Making use of https://github.com/galaxyproject/galaxy/pull/117
- [ ] liftOver_wrapper has disabled test cases. https://github.com/galaxyproject/galaxy/commit/77b167bbd1d6d466aa7caeeafbc9e494d7e480b8
- [ ] Historical behaviour question on [this](https://github.com/galaxyproject/galaxy/blob/77b167bbd1d6d466aa7caeeafbc9e494d7e480b8/tools/extract/liftOver_wrapper.xml#L49), does `format="input"` imply `format_source="first_input_dataset"`?
- [ ] how do we feel about test-data again? Some tools are easy to add tests for (e.g. commWrapper.xml/pl)
- [ ] should `joiner2.xml` be removed? when `joiner.xml` exists?
- [x] command line parameters in maf_\* were re-arranged as part of the rewrite to de-duplicate CLI strings.
- [x] is [this](https://github.com/galaxyproject/galaxy/blob/dev/tools/meme/fimo.xml#L39) intentional?
- [ ] could someone provide documentation for https://github.com/galaxyproject/galaxy/blob/dev/tools/validation/fix_errors.xml ?

Related TODOs
- [ ] https://github.com/galaxyproject/planemo/issues/233
- [ ] https://github.com/galaxyproject/planemo/issues/232
- [ ] https://github.com/galaxyproject/planemo/issues/231

### CODE REVIEW

**Comment by jmchilton** [comment]
I don't know if it would work but I feel like the data source tools should essentially all have citations. The data source tools should assign a tool id to the outputs and those ids will resolve to tools in the toolbox - so I feel like say building a citation list for a history should work for these outputs. At very least I wouldn't add empty citation tags to please planemo thereby making it more likely we would forget to add citations and make them work in the future.

**Comment by jmchilton** [comment]
I do like adding help text though - because this will probably come through in the API for instance.

**Comment by nsoranzo** [comment]
> is [this](https://github.com/galaxyproject/galaxy/blob/dev/tools/meme/fimo.xml#L39) intentional?

I think so, the `'` is opened at https://github.com/galaxyproject/galaxy/blob/dev/tools/meme/fimo.xml#L4 .

**Comment by hexylena** [comment]
@nsoranzo great, thanks. I couldn't seem to see where that was.

@jmchilton absolutely makes sense. Should they cite the "Blankenberg et al. in preparation" or whatever that citation was that got applied to so many tools?

**Comment by martenson** [comment]
this is great, especially proper exit code detection is a tremendous improvement

**Comment by jmchilton** [comment]
@erasche - I guess every data source in core should have a citation for 10.1093/database/bar011 - (a Blankenberg, et. al. out of prep :)) and then a citation for the underlying tool. So I guess
- 10.1101/gr.229102 for UCSC genome browser
- 10.1093/nar/gkv350 for biomart
  etc...

File: tools/visualization/LAJ.xml [hunk 1]
```diff
@@ -28,5 +31,5 @@ Laj is a tool for viewing and manipulating the output from pairwise alignment pr
 **Note:** If you save output from the applet, you will need to manually refresh your history.
 
   </help>
-  <code file="LAJ_code.py"/>
+  <citations />
```
Comment (jmchilton) [inline tools/visualization/LAJ.xml:34]: <citation type="bibtex">
@misc{Miller2005,
  author = {Miller Lab},
  year = {2005},
  title = {Laj},
  url = {http://globin.bx.psu.edu/dist/laj/},
}
</citation>

File: tools/stats/gsummary.xml.groups [hunk 2]
```diff
@@ -52,11 +53,12 @@ This tool computes basic summary statistics on a given column, or on an expressi
...
-**TIP:** Most functions (like *abs*) take only a single expression. *log* can take one or two parameters, like *log(expression,base)* 
+**TIP:** Most functions (like *abs*) take only a single expression. *log* can take one or two parameters, like *log(expression,base)*
 
 Currently, these R functions are supported: *abs, sign, sqrt, floor, ceiling, trunc, round, signif, exp, log, cos, sin, tan, acos, asin, atan, cosh, sinh, tanh, acosh, asinh, atanh, lgamma, gamma, gammaCody, digamma, trigamma, cumsum, cumprod, cummax, cummin*
 
 .. |INFO| image:: ./static/images/icon_info_sml.gif
 
-</help>
+  </help>
+  <citations />
```
Comment (jmchilton) [inline tools/stats/gsummary.xml.groups:63]: I wouldn't add the empty citations block. @blankenberg do you think this tool is correlated with any particular Galaxy paper?

File: tools/sr_mapping/srma_wrapper.xml [hunk 3]
```diff
@@ -181,38 +182,39 @@ This is an exhaustive list of SRMA options:
...
+                                default value. Possible values: {true, false}
+
+  USE_SEQUENCE_QUALITIES=BooleanUse sequence qualities Default value: true. This option can be set to 'null' to clear the
+                                default value. Possible values: {true, false}
+
+  MAX_HEAP_SIZE=Integer         The maximum number of nodes on the heap before re-alignment is ignored Default value:
+                                8192. This option can be set to 'null' to clear the default value.
 
   </help>
+  <citations />
```
Comment (jmchilton) [inline tools/sr_mapping/srma_wrapper.xml:219]: 10.1093/bioinformatics/bts286

File: tools/sr_mapping/mosaik.xml [hunk 4]
```diff
@@ -123,4 +123,5 @@
   <help>
 This tool uses Mosaik to align reads to a reference sequence.
   </help>
+  <citations />
```
Comment (jmchilton) [inline tools/sr_mapping/mosaik.xml:126]: 10.1371/journal.pone.0090581

File: tools/sr_mapping/fastq_statistics.xml [hunk 5]
```diff
@@ -83,7 +85,8 @@ Creates quality statistics report for the given Solexa/FASTQ library.
     34	6362991	-5	40	140638447	22.10	10	20	40	30	-5	40	1501615	1637357	18113	3205237	669
     35	6362991	-5	40	138910532	21.83	10	20	40	30	-5	40	1532519	3495057	23229	1311834	352
     36	6362991	-5	40	117158566	18.41	7	15	30	23	-5	40	4074444	1402980	63287	822035	245
-    
 
-</help>
+
+  </help>
+  <citations />
```
Comment (jmchilton) [inline tools/sr_mapping/fastq_statistics.xml:91]: Probably this -> 10.1093/bioinformatics/btq281

File: tools/sr_mapping/bfast_wrapper.xml [hunk 6]
```diff
@@ -322,58 +373,9 @@ For **postprocess**::
...
-    <test>
-      <param name="input1" ftype="fastqsanger" value="random_phiX_1.fastqsanger" />
-      <param name="refGenomeSource_type" value="indexed" />
-      <param name="indices" value="phiX_nt_50" />
-      <param name="source_select" value="pre_set" />
-      <param name="suppressHeader" value="" />
-      <output name="output" ftype="sam" file="bfast_out3.sam" lines_diff="2" /><!-- MD:Z:11T38 instead of MD:Z:50 on one line-->
-    </test>
-  </tests>
+  <citations/>
```
Comment (jmchilton) [inline tools/sr_mapping/bfast_wrapper.xml:380]: 10.1371/journal.pone.0007767

**Comment by hexylena** [comment]
@jmchilton amazing, doing my citation lookup work for me, much appreciated!!

**Comment by jmchilton** [comment]
No - we are both doing @blankenberg's work for him :) - what do we get out of this.

There are a ton though - so unless you want to track everyone down - I would just eliminate the empty citations blocks and run planemo with that check excluded. We can come back to this some other day. If you are in the mood to track them all down though - I won't complain :).

**Comment by hexylena** [comment]
Agreed, I'll remove the citation blocks so we can keep generating warnings until that date

(I had hoped that we'd be able to add planemo linting to the travis job, but without citations for everything, without test cases for everything, there's no sense in turning linting on)

**Comment by jmchilton** [comment]
I understand the desire to add those checks to travis - though I don't think we will be modifying these tools frequently so it is probably better to spend our Travis time on something like including some integration testing or API tests.

**Comment by hexylena** [comment]
Eh, fair point. I just want to test ALL the things :P

**Comment by jmchilton** [comment]
> Historical behaviour question on this, does format="input" imply format_source="first_input_dataset"?

If I recall correctly, I actually think `format="input"` implies `format_source="random_input_dataset"` - since this stuff is set by iterating over dict and just keeping the variable set to the last element encountered. This is why planemo discourages it.

**Comment by hexylena** [comment]
Gotcha, that was the assumption I was working under. Thanks for the confirmation.

I replaced a lot of `format` and `format_source` with fixed datatypes, because most of the time that was used, the input was hardcoded to tabular.

**Comment by jmchilton** [comment]
I want to test all the things also - but IMO travis is only a place to test the most usefully and quickly tested things.

**Comment by hexylena** [comment]
Yep, fair enough. Wish travis had a better concept of "what's changed since last time", and it didn't require ugly hacks in Jenkins to do those sort of tests.

**Comment by jmchilton** [comment]
> how do we feel about test-data again? Some tools are easy to add tests for (e.g. commWrapper.xml/pl)

Would love to see a tests added - just commit the test-data to https://github.com/galaxyproject/galaxy-test-data instead of here. It will be fetched dynamically as needed by the test framework.

**Comment by jmchilton** [comment]
> I replaced a lot of format and format_source with fixed datatypes, because most of the time that was used, the input was hardcoded to tabular.

That is probably a mistake - many tabular tools when operating a bed file for instance should produce another bed file. The inputs are polymorphic - and so the output datatypes should be inherited instead of hardcoded.

**Comment by hexylena** [comment]
.... eww. That's...wow, okay. I didn't remember that bed datatypes subclassed tabular and could be accepted as inputs there. The rest of the format tree is so flat...

**Comment by hexylena** [comment]
Note to self from #524:

> Note separately that if the first dataset is considered as a bed file, then the output file is also considered bed, and this appears to wrongly mark column c6 (description) as strand.

So whenever I go back to refactor the datatypes I mistakenly hardcoded to tabular, there are cases when `format_source` isn't appropriate.

**Comment by jmchilton** [comment]
New oldest PR, I'm thinking i going to try to unwind into smaller PRs and then convert what is left into an issue - is that okay @erasche or were you super eager to return to this mega effort?

**Comment by jmchilton** [comment]
xref #846

**Comment by hexylena** [comment]
@jmchilton you are absolutely welcome to. I may likewise try and find time to unwind a task or two out of this. Thanks for reviving this zombie :)

**Comment by jmchilton** [comment]
Okay - I have transitioned all my DOI comments into PRs. Maybe it would be best to just remove all the empty citations blocks in this PR.

**Comment by hexylena** [comment]
@jmchilton it may be best to just kill this PR and use the smaller ones you've been opening. This PR did a couple things very wrong (i.e. removing `format_from` statements)

**Comment by jmchilton** [comment]
@erasche - there is good stuff in there, but yeah I agree lets just kill it for now and try to pull out the good stuff in small doses going forward.

**Pull request closed without merging**